### PR TITLE
Introduce `bal openapi sanitize` sub command

### PR DESCRIPTION
**Pull request by TharmiganK**
Part of: https://github.com/ballerina-platform/ballerina-library/issues/6977

This modifies all the schema names with Ballerina friendly names. The following schemas can not be modified directly:
- Query and Header Parameters
- Properties of Object schemas

In that case a Ballerina extension: `x-ballerina-name` is added to the schemas, which will be used to generate annotations in the client/service generation. Refer to this [proposal](https://github.com/ballerina-platform/ballerina-library/issues/6867) for more details.

## Description

```console
NAME
       bal openapi sanitize - Sanitize the OpenAPI contract file according to
       the best practices of Ballerina.

SYNOPSIS
       bal openapi sanitize [-i | --input] <openapi-contract-file-path>
                            [-o | --output] <output-file-path>
                            [-n | --name] <generated-file-name>
                            [-f | --format] [json|yaml]
                            [-t | --tags] <tag-names>
                            [--operations] <operation-names>

DESCRIPTION
       Sanitize the OpenAPI contract file according to the best naming
       practices of Ballerina. The OpenAPI contract is flatten and the type
       schema names are made Ballerina friendly. The Ballerina name
       extensions are added to the schemas which can not be modified
       directly.


OPTIONS
       -i, --input <openapi-contract-file-path>
            This is a mandatory input. The given OpenAPI contract will be sanitized.
            The OpenAPI contract can be either a YAML or a JSON.

       -o, --output <output-file-path>
            This is an optional input. The given output file path will be used to
            save the sanitized OpenAPI contract. The default output file path is
            the executed directory.

       -n, --name <generated-file-name>
            This is an optional input. The given name will be used to save the
            sanitized OpenAPI contract. The default name is `sanitized_openapi`.

       -f, --format [json|yaml]
            This is an optional input. The sanitized OpenAPI contract will be
            saved in the given format. The format can be either JSON or YAML.
            The default format is same as the input file format.

       -t, --tags <tag-names>
            This is an optional input. The sanitized OpenAPI contract will only
            have the operations with the given tags.

       --operations <operation-names>
            This is an optional input. The sanitized OpenAPI contract will only
            have the given operations.

EXAMPLES
       Sanitize the `service.yaml` OpenAPI contract file.
           $ bal openapi sanitize -i service.yaml

       Sanitize the `service.yaml` OpenAPI contract file and save it as
       `sanitized_svc.json` file.
           $ bal openapi sanitize -i hello.yaml -n sanitized_svc -f json

       Sanitize the `service.json` OpenAPI contract file by filtering the
       operations with the `service` tag.
           $ bal openapi sanitize -i service.json -t service
```

### CODE REVIEW

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/Sanitize.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+            if (!format.equalsIgnoreCase("json") && !format.equalsIgnoreCase("yaml")) {
+                setDefaultFormat();
+                errorStream.println(WARNING_INVALID_OUTPUT_FORMAT);
+            }
+        } else {
+            setDefaultFormat();
+        }
+
+        if (Objects.isNull(fileName)) {
+            fileName = "sanitized_openapi";
```
[the remark this answers is not in the document]
Comment (TharmiganK) [reply openapi-cli/src/main/java/io/ballerina/openapi/cmd/Sanitize.java:157]: Sure

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/common/OASModifier.java [hunk 2]
```diff
@@ -106,20 +106,42 @@ private static Map<String, String> getResolvedNameMapping(Map<String, String> na
...
+        if (Objects.isNull(components) || Objects.isNull(components.getSchemas())) {
+            return;
+        }
+
+        Map<String, Schema> schemas = components.getSchemas();
+        for (Map.Entry<String, Schema> schema : schemas.entrySet()) {
+            Schema schemaValue = schema.getValue();
+            if (schemaValue instanceof ObjectSchema objectSchema) {
+                Map<String, Schema> properties = objectSchema.getProperties();
+                if (properties != null && !properties.isEmpty()) {
```
Comment (ayeshLK) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/common/OASModifier.java:126]:
```suggestion
                if (Objects.nonNull(properties) && !properties.isEmpty()) {
```

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/common/OASModifier.java [hunk 3]
```diff
@@ -268,19 +290,67 @@ private static String updateParameters(String pathValue, Map<String, String> par
...
         for (Parameter parameter : parameters) {
-            if (!parameter.getIn().equals("path")) {
-                modifiedParameters.add(parameter);
-                continue;
+            String location = parameter.getIn();
+            if ("path".equals(location)) {
+                String modifiedPathParam = parameterNames.get(parameter.getName());
+                parameter.setName(modifiedPathParam);
+                pathValue = replaceContentInBraces(pathValue, modifiedPathParam);
+            } else if ("query".equals(location) || "header".equals(location)) {
```
Comment (ayeshLK) [inline openapi-core/src/main/java/io/ballerina/openapi/core/generators/common/OASModifier.java:298]: IMO we should have predefined constants for `path`, `query`, and `header` references.

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/Sanitize.java [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+    private String getOutputFileName() {
+        return resolveContractFileName(targetPath, fileName + getFileExtension(), format.equals("json"));
+    }
+
+    private String getFileExtension() {
+        return (Objects.nonNull(format) && format.equals("json")) ? JSON_EXTENSION : YAML_EXTENSION;
+    }
+
+    private void filterOpenAPIOperations(OpenAPI openAPI) {
```
Comment (lnash94) [inline openapi-cli/src/main/java/io/ballerina/openapi/cmd/Sanitize.java:238]: Can we reuse the method `filterOpenAPIOperations` for this ?

Comment (TharmiganK) [reply openapi-cli/src/main/java/io/ballerina/openapi/cmd/Sanitize.java:238]: Yes, there are common utils between this and the flatten sub command implementation. Will reuse them when merging one with another

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/common/OASModifier.java (hunk 3 above)
Comment (TharmiganK) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/common/OASModifier.java]: Fixed via [c0360b9](https://github.com/ballerina-platform/openapi-tools/pull/1770/commits/c0360b9e7086ebe02f649d1b10c3a2b34f392971)

File: openapi-core/src/main/java/io/ballerina/openapi/core/generators/common/OASModifier.java (hunk 2 above)
Comment (TharmiganK) [reply openapi-core/src/main/java/io/ballerina/openapi/core/generators/common/OASModifier.java]: Fixed via [c0360b9](https://github.com/ballerina-platform/openapi-tools/pull/1770/commits/c0360b9e7086ebe02f649d1b10c3a2b34f392971)

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/Sanitize.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (TharmiganK) [reply openapi-cli/src/main/java/io/ballerina/openapi/cmd/Sanitize.java]: Fixed via [c0360b9](https://github.com/ballerina-platform/openapi-tools/pull/1770/commits/c0360b9e7086ebe02f649d1b10c3a2b34f392971)

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/Sanitize.java (hunk 4 above)
Comment (TharmiganK) [reply openapi-cli/src/main/java/io/ballerina/openapi/cmd/Sanitize.java]: Fixed via [0c67edd](https://github.com/ballerina-platform/openapi-tools/pull/1770/commits/0c67edd6cfb2855ca539beb0e06a9300d88d3fb8)

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/SubCmdBase.java [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        }
+
+        if (Objects.nonNull(outputPath)) {
+            targetPath = Paths.get(outputPath).isAbsolute() ?
+                    Paths.get(outputPath) : Paths.get(targetPath.toString(), outputPath);
+        }
+    }
+
+    private void setDefaultFormat() {
+        format = inputPath.endsWith(JSON_EXTENSION) ? JSON : "yaml";
```
Comment (lnash94) [inline openapi-cli/src/main/java/io/ballerina/openapi/cmd/SubCmdBase.java:262]: Shall we use constant for `yaml`?

File: openapi-cli/src/test/java/io/ballerina/openapi/cmd/NegativeCmdTests.java [hunk 6]
```diff
@@ -95,6 +95,61 @@ public void testFlattenWithInputOpenAPIFileParsingIssues() throws IOException {
...
+
+    @Test(description = "Test with the input OpenAPI file in `sanitize` sub command which has parsing issues")
+    public void testSanitizeWithInputOpenAPIFileParsingIssues() throws IOException {
+        String[] args = {"-i", resourceDir + "/cmd/sanitize/openapi_invalid.json", "-f", "txt", "-o",
+                tmpDir.toString()};
+        Sanitize sanitize = new Sanitize(printStream, false);
+        new CommandLine(sanitize).parseArgs(args);
+        sanitize.execute();
+        String output = readOutput(true);
+        Assert.assertTrue(output.contains("WARNING: invalid output format. The output format should be either " +
```
Comment (lnash94) [inline openapi-cli/src/test/java/io/ballerina/openapi/cmd/NegativeCmdTests.java:152]: If the user provides openapi with parser issues, are we continuing the sanitaization or terminate by providing encountered issue list?

File: openapi-cli/src/test/java/io/ballerina/openapi/cmd/OpenAPICmdTest.java [hunk 7]
```diff
@@ -1025,6 +1025,89 @@ public void testFlattenCmdWithComposedSchema() throws IOException {
...
+    @Test(description = "Test openapi sanitize sub command with composed schema")
+    public void testSanitizeCmdWithComposedSchema() throws IOException {
+        Path expectedFilePath = resourceDir.resolve(Paths.get("cmd/sanitize/sanitized_openapi_composed_schema.json"));
+        String[] args = {"-i", resourceDir + "/cmd/sanitize/openapi_composed_schema.json", "-o", tmpDir.toString()};
+        Sanitize sanitize = new Sanitize();
+        new CommandLine(sanitize).parseArgs(args);
+        sanitize.execute();
+        compareFiles(expectedFilePath, tmpDir.resolve("sanitized_openapi.json"));
+    }
+
```
Comment (lnash94) [inline openapi-cli/src/test/java/io/ballerina/openapi/cmd/OpenAPICmdTest.java:1110]: Shall we add tests for different OpenAPI versions, (v2, v3.0.0, v3.0.1)?

File: openapi-cli/src/test/resources/cmd/sanitize/openapi_composed_schema.json [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+      },
+      "album": {
+        "required": [
+          "_id",
+          "artist",
+          "title"
+        ],
+        "type": "object",
+        "properties": {
+          "_id": {
```
Comment (lnash94) [inline openapi-cli/src/test/resources/cmd/sanitize/openapi_composed_schema.json:303]: Shall we have a test something with `\`?

File: openapi-cli/src/test/java/io/ballerina/openapi/cmd/NegativeCmdTests.java (hunk 6 above)
Comment (TharmiganK) [reply openapi-cli/src/test/java/io/ballerina/openapi/cmd/NegativeCmdTests.java:152]: We are terminating if we can't get the model from the parser. If there are only warnings, we continue.

File: openapi-cli/src/test/java/io/ballerina/openapi/cmd/OpenAPICmdTest.java (hunk 7 above)
Comment (TharmiganK) [reply openapi-cli/src/test/java/io/ballerina/openapi/cmd/OpenAPICmdTest.java:1130]: Fixed via [4befa98](https://github.com/ballerina-platform/openapi-tools/pull/1770/commits/4befa98e26f9a3af611f1cce7327d236ff7bf30e)

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/SubCmdBase.java (hunk 5 above)
Comment (TharmiganK) [reply openapi-cli/src/main/java/io/ballerina/openapi/cmd/SubCmdBase.java]: Fixed via [ad24c75](https://github.com/ballerina-platform/openapi-tools/pull/1770/commits/ad24c758c60adc0ffb60cdb0461b955c5994e913)

File: openapi-cli/src/test/resources/cmd/sanitize/openapi_composed_schema.json (hunk 8 above)
Comment (TharmiganK) [reply openapi-cli/src/test/resources/cmd/sanitize/openapi_composed_schema.json:303]: Fixed via [49b5e31](https://github.com/ballerina-platform/openapi-tools/pull/1770/commits/49b5e31e836e1800b9cafc85a7e6d24d393a5f01)

**Pull request merged**